improver/regrid/grid.py

```diff
@@ -0,0 +1,393 @@
+    nearest_is_land = in_is_land[nearest_in_indexes]  # dimensions M x K
+
+    # these input points surrounding output points have the same surface type
+    nearest_same_type = np.logical_not(
+        np.logical_xor(nearest_is_land, out_is_land_bcast)
+    )  # dimensions M x K
+
+    return nearest_same_type
+
+
+def slice_cube_by_domain(cube_in, output_domain):
+    """
+    extract cube domain to be consistent as cube_reference's domain
+
+    Args:
+        cube_in (iris.cube.Cube):
+            input data cube to be sliced
+        output_domain(tuple):
+            lat_max,lon_max,lat_min,lon_min
+
+    Returns:
+        cube_in (iris.cube.Cube):
+            data cube after slicing
+    """
+    lat_max, lon_max, lat_min, lon_min = output_domain
+    lat_d, lon_d = get_grid_spacing(cube_in)
+
+    domain = iris.Constraint(
+        latitude=lambda val: lat_min - 2.0 * lat_d < val < lat_max + 2.0 * lat_d
+    ) & iris.Constraint(
+        longitude=lambda val: lon_min - 2.0 * lon_d < val < lon_max + 2.0 * lon_d
+    )
+
+    cube_in = cube_in.extract(domain)
+
+    return cube_in
+
+
+def slice_mask_cube_by_domain(cube_in, cube_in_mask, output_domain):
+    """
+    extract cube domain to be consistent as cube_reference's domain
+
+    Args:
+        cube_in (iris.cube.Cube):
+            input data cube to be sliced
+        cube_in_mask (iris.cube.Cube):
+            input maskcube to be sliced
+        output_domain (Tuple[float, float, float, float]):
+            lat_max, lon_max, lat_min, lon_min
+    Returns:
+        Tuple[iris.cube.cube, iris.cube.Cube]:
+            data cube after slicing, mask cube after slicing
+    """
+    lat_max, lon_max, lat_min, lon_min = output_domain
+    lat_d_1, lon_d_1 = get_grid_spacing(cube_in)
+    lat_d_2, lon_d_2 = get_grid_spacing(cube_in_mask)
+    lat_d = lat_d_1 if lat_d_1 > lat_d_2 else lat_d_2
+    lon_d = lon_d_1 if lon_d_1 > lon_d_2 else lon_d_2
+
+    domain = iris.Constraint(
+        latitude=lambda val: lat_min - 2.0 * lat_d < val < lat_max + 2.0 * lat_d
+    ) & iris.Constraint(
+        longitude=lambda val: lon_min - 2.0 * lon_d < val < lon_max + 2.0 * lon_d
+    )
+
+    cube_in = cube_in.extract(domain)
+    cube_in_mask = cube_in_mask.extract(domain)
+
+    return cube_in, cube_in_mask
+
+
+def create_regrid_cube(cube_array, cube_in, cube_out):
+    """
+    create a regridded cube from regridded value(numpy array)
+    source cube cube_in must be in  GeogCS's lats/lons system
+    tergat cube_out either lats/lons system or LambertAzimuthalEqualArea system
+
+    Args:
+        cube_array (numpy ndarray):
+            regridded value (multidimensional)
+        cube_in (iris.cube.Cube):
+            source cube (for value's non-grid dimensions and attributes)
+        cube_out (iris.cube.Cube):
+            target cube (for target grid information)
+
+    Returns:
+         iris.cube.Cube: regridded result cube
+    """
+    cube_coord_names = get_cube_coord_names(cube_in)
```

Comment:
I wonder if this code could be simpler if you find all your cube coordinates you want to copy across, add on the new x and y coordinates and then build your cube all at once.

Here's some half written code for you to adapt if you wish.

```  
  aux_coords_and_dims, dim_coords_and_dims = [
        [
            (coord.copy(), template_cube.coord_dims(coord))
            for coord in getattr(template_cube, coord_type) if coord not lat or lon
        ]
        for coord_type in ("aux_coords", "dim_coords")
    ]

Append new x and y coords with corresponding dims to the coord lists.

    cube = iris.cube.Cube(
        data,
        units=cube_in.units,
        attributes=cube_in.attributes,
        ...
        dim_coords_and_dims=dim_coords_and_dims,
        aux_coords_and_dims=aux_coords_and_dims,
    )
    cube.rename(name)

    return cube
```

Reply:
I will try it.